torch/quantization/observer.py

```diff
@@ -799,6 +802,129 @@ def forward(self, x_orig):
         return x_orig
 
 
+class InputWeightMinMaxObserver(PerChannelMinMaxObserver):
+    r"""Observer module for computing the scale factor needed for input-weight
+    equalization based on the running min and max values of the input and weight
+    columns.
+
+    This observer uses the tensor min/max statistics to compute the column
+    quantization parameters. The module records the running minimum and maximum
+    of incoming tensors, and uses this statistic to compute the scale factor.
+
+    Args:
+        ch_axis: Channel axis
+        dtype: Quantized data type
+        qscheme: Quantization scheme to be used
+        reduce_range: Reduces the range of the quantized data type by 1 bit
+        quant_min: Minimum quantization value. If unspecified, it will follow the 8-bit setup.
+        quant_max: Maximum quantization value. If unspecified, it will follow the 8-bit setup.
+
+    Given running min/max of the input columns as :math:`x_\text{min}` and :math:`x_\text{max}`,
+    and the running min/max of the weight columns as :math:`w_\text{min}` and :math:`w_\text{max}`,
+    scale :math:`S` is computed as:
+
+    The running minimum/maximum :math:`x_\text{min/max}` and
+    :math:`w_\text{min/max}` are computed in the same way as
+    :class:`~torch.quantization.observer.MinMaxObserver`, with the difference
+    that the running min/max values are stored per column.
+
+    The scale factor :math:`S` is then computed as:
+
+    .. math::
+        S = \sqrt{\frac{x_{max} - x_{min}}{w_{max} - w{min}}}
+
+    where :math:`X` is the observed tensor.
+
+    .. note:: If the running minimum equals to the running maximum, the scales
+              and zero_points are set to 1.0 and 0.
+    """
+    min_inputs_col: torch.Tensor
+    max_inputs_col: torch.Tensor
+    min_weights_col: torch.Tensor
+    max_weights_col: torch.Tensor
+    min_weights_row: torch.Tensor
+    max_weights_row: torch.Tensor
+
+    def __init__(self, dtype=torch.quint8,
+                 qscheme=torch.per_channel_affine, reduce_range=False,
+                 quant_min=None, quant_max=None, factory_kwargs=None) -> None:
+        super(InputWeightMinMaxObserver, self).__init__(ch_axis=1, dtype=dtype,
+                                                        qscheme=qscheme,
```

Comment:
I think qscheme is confusing here because we don't know if it corresponds to input or weight. Can we distinguish it? In the code we can assert that we only expect per-tensor for input.